stl/inc/xutility

```diff
@@ -4650,83 +4650,90 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
 #endif // _HAS_CXX17
 
 // FUNCTION TEMPLATE equal
-// Tests whether the usual arithmetic conversions will preserve the bit-pattern when promoting to int
-// e.g. short == unsigned short -> false
-//      int == unsigned int -> true
-template <class _Elem1, class _Elem2>
-_INLINE_VAR constexpr bool _Value_equality_is_bitwise_equality = static_cast<_Elem1>(-1) == static_cast<_Elem2>(-1);
-
-// determines whether it is safe to call memcmp to compare things
-template <class _Elem1, class _Elem2, class _Pr>
-_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper = false;
 
-// allow memcmping pointers-to-cv-T with equal_to<>
-template <class _Elem1, class _Elem2>
-_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>> =
-    is_same_v<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>;
+// _Can_memcmp_elements<_Elem1, _Elem2> reports whether `_Elem1 == _Elem2` can be optimized to memcmp.
+// Here, _Elem1 and _Elem2 aren't top-level const, because we remove_const_t before using _Can_memcmp_elements.
+
+// Integral types are eligible for memcmp in very specific cases.
+// * They must be the same size. (`int == long` is eligible; `int == long long` isn't.)

```

Comment:
I wonder whether you could add std::byte directly here. That would mean expanding on _Is_nonboool_integral. 

This would allow comparing std::byte with integers of compatible size. If this is the right idea I do not know

Reply:
That would be an excellent idea if `std::byte` were equality comparable with integers. However, it was designed to prohibit such comparisons. It's an `enum class` which lacks the implicit conversions of a plain `enum`:
https://github.com/microsoft/STL/blob/212de15c590010c961c0e7bb4025f6fec89866c8/stl/inc/cstddef#L29
It's convertible to integers via a named function `to_integer`:
https://github.com/microsoft/STL/blob/212de15c590010c961c0e7bb4025f6fec89866c8/stl/inc/cstddef#L87-L90
But `byte == integer` is a compiler error, so we must not bypass that by calling `memcmp`.

Reply:
Derp, you are indeed correct. It is kind of sad that the user cannot really project to integer ranges though as we would reject the comparator

Reply:
Yeah - at that point, we basically need to ask the backend to perform this optimization. (My historical understanding is that this metaprogramming was very important when the backend couldn't recognize raw loops and optimize them to `memcmp`; nowadays the metaprogramming mostly benefits non-optimized debug mode.)